Assets/Wizards Code/MaxArchxd/Scripts/DangerousAreaScanner.cs

```diff
@@ -0,0 +1,141 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.AI;
+using WizardsCode.BackgroundAI;
+
+[RequireComponent(typeof(NavMeshAgent))]
+public class DangerousAreaScanner : AbstractSingleton<DangerousAreaScanner>
+{
+    [Tooltip("The agent will scan for danger before entering an area of the NavMesh on this list")]
+    public List<string> dangerousAreas;
+    [Tooltip("The agent will scan for obstacles with these tags. Leave the list empty to scan for all obstacles")]
+    public List<string> dangerousObstacles;
+    [Tooltip("Amount of time in seconds the agent will wait in place before scanning again")]
+    public float waitTime = 3;
+    [Tooltip("Amount of time in seconds between scans while the agent is moving")]
+    public float scanInterval = 1;
+    [Tooltip("Radius in which the agent will scan for danger")]
+    public int detectionRadius = 10;
+
+    private NavMeshAgent agent;
+    private HashSet<int> dangerousAreasLayers;
+    private int currentArea;
+    private bool waiting;
+    private float timeSinceLastScan;
+
+    private void Awake()
+    {
+        agent = GetComponent<NavMeshAgent>();
+        dangerousAreasLayers = new HashSet<int>();
+        timeSinceLastScan = scanInterval;
+    }
+
+    void Start()
+    {
+        updateDangerousAreas();
+        currentArea = getCurrentArea();
+    }
+
+    public void updateDangerousAreas()
+    {
+        foreach (string layerName in dangerousAreas)
+        {
+            int areaIndex = NavMesh.GetAreaFromName(layerName);
+            if(areaIndex != -1) dangerousAreasLayers.Add(1 << areaIndex);
+        }
+    }
+
+    void Update()
+    {
+        if (waiting) return;
+        timeSinceLastScan += Time.deltaTime;
```

Comment:
Should be running as a Coroutine rather than an update as this will only run every 1s. This will also remove the need to track timeSinceLastScan.